libraries/AP_GPS/AP_GPS_SBP2.cpp

```diff
@@ -0,0 +1,494 @@
+                if (parser_state.n_read >= 2) {
+                    parser_state.n_read = 0;
+                    parser_state.state = sbp_parser_state_t::GET_SENDER;
+                }
+                break;
+
+            case sbp_parser_state_t::GET_SENDER:
+                *((uint8_t*)&(parser_state.sender_id) + parser_state.n_read) = temp;
+                parser_state.n_read += 1;
+                if (parser_state.n_read >= 2) {
+                    parser_state.n_read = 0;
+                    parser_state.state = sbp_parser_state_t::GET_LEN;
+                }
+                break;
+
+            case sbp_parser_state_t::GET_LEN:
+                parser_state.msg_len = temp;
+                parser_state.n_read = 0;
+                parser_state.state = sbp_parser_state_t::GET_MSG;
+                break;
+
+            case sbp_parser_state_t::GET_MSG:
+                *((uint8_t*)&(parser_state.msg_buff) + parser_state.n_read) = temp;
+                parser_state.n_read += 1;
+                if (parser_state.n_read >= parser_state.msg_len) {
+                    parser_state.n_read = 0;
+                    parser_state.state = sbp_parser_state_t::GET_CRC;
+                }
+                break;
+
+            case sbp_parser_state_t::GET_CRC:
+                *((uint8_t*)&(parser_state.crc) + parser_state.n_read) = temp;
+                parser_state.n_read += 1;
+                if (parser_state.n_read >= 2) {
+                    parser_state.state = sbp_parser_state_t::WAITING;
+
+                    crc = crc16_ccitt((uint8_t*)&(parser_state.msg_type), 2, 0);
+                    crc = crc16_ccitt((uint8_t*)&(parser_state.sender_id), 2, crc);
+                    crc = crc16_ccitt(&(parser_state.msg_len), 1, crc);
+                    crc = crc16_ccitt(parser_state.msg_buff, parser_state.msg_len, crc);
+                    if (parser_state.crc == crc) {
+                        _sbp_process_message();
+                    } else {
+                        Debug("CRC Error Occurred!");
+                        crc_error_counter += 1;
+                    }
+
+                    parser_state.state = sbp_parser_state_t::WAITING;
+                }
+                break;
+
+            default:
+                parser_state.state = sbp_parser_state_t::WAITING;
+                break;
+            }
+    }
+}
+
+
+//INVARIANT: A fully received message with correct CRC is currently in parser_state
+void
+AP_GPS_SBP2::_sbp_process_message() {
+
+    //Here, we copy messages into local structs.
+    switch (parser_state.msg_type) {
+        case SBP_HEARTBEAT_MSGTYPE:
+            last_heartbeat = *((struct sbp_heartbeat_t*)parser_state.msg_buff);
+            last_heartbeat_received_ms = AP_HAL::millis();
+            waiting_for_first_heartbeat = false;
+            break;
+
+        case SBP_GPS_TIME_MSGTYPE:
+            last_gps_time = *((struct sbp_gps_time_t*)parser_state.msg_buff);
```

Comment:
This is a bit confusing that the tow is just the first value in the buffer, could the buffer instead be presented as a union of the message types + array, that way this addressing could instead be parser_state.msg_buff.gps_time.tow